gobblin-core/src/main/java/gobblin/source/extractor/extract/sftp/SftpFsHelper.java

```diff
@@ -51,10 +55,52 @@ public SftpFsHelper(State state) {
     this.state = state;
   }
 
-  public ChannelSftp getSftpConnection() {
-    return this.channelSftp;
+  /**
+   * Create new channel every time a command needs to be executed. This is required to support execution of multiple
+   * commands in parallel. All created channels are cleaned up when the session is closed.
+   * @see #cleanupSftpChannels()
+   * @see #close()
+   *
+   * @return a new {@link ChannelSftp}
+   * @throws SftpException
+   */
+  public ChannelSftp getSftpChannel() throws SftpException {
+    ChannelSftp channelSftp;
+    try {
+      channelSftp = (ChannelSftp) session.openChannel("sftp");
+      channelSftp.connect();
+      channels.add(channelSftp);
+      return channelSftp;
+    } catch (JSchException e) {
+      throw new SftpException(0, "Cannot open a channel to SFTP server", e);
+    }
+  }
+
+  public ChannelExec getExecChannel(String command) throws SftpException {
+    ChannelExec channelExec;
+    try {
+      channelExec = (ChannelExec) session.openChannel("exec");
+      channelExec.setCommand(command);
+      channelExec.connect();
+      channels.add(channelExec);
+      return channelExec;
+    } catch (JSchException e) {
+      throw new SftpException(0, "Cannot open a channel to SFTP server", e);
+    }
   }
 
+  private void cleanupSftpChannels(){
+    log.info("Exiting all open channels");
+    for (Channel channel : channels) {
+      if (channel != null) {
+        if (channel instanceof ChannelSftp) {
+          ((ChannelSftp)channel).exit();
```

Comment:
1. The javadocs say "exit" is a synonym of disconnect() so you probably don't need the cast.
2. I'd put the exit()/disconnect() in a try-catch (RuntimeException) block so that errors in on channel do not affect the others. If there was an exception thrown, you can record it and re-throw it at the end.